### DISCUSSION TITLE
Pay::UserMailer#receipt job failing

### DISCUSSION THREAD

**Discussion by EsotericCoder** [pay-rails/pay#636] [Q&A]
Was able to set up payments and successfully receiving charges. Pay::UserMailer#receipt job keeps failing since KeyError: key not found: :details. Did not see any configuration for details in the set up guide. Is there something I am missing?

**Comment by excid3** [comment]
Please provide some more details like the stacktrace.

**Reply by EsotericCoder** [reply]
I am redirect checkout to stripe instead of handling it on site. Everything works correctly, new charges are being saved to the database and status is updating paid.

`@checkout_session = current_user.payment_processor.checkout(
        mode: "payment",
        line_items: @lesson.stripe_price_id,
        success_url: "#{request.base_url}",
        cancel_url: "#{request.base_url}"
      )
      redirect_to @checkout_session.url, allow_other_host: true`

The error occurs in Sidekiq as it tries to run the MailDeliverJob for receipts

`2022-06-12T23:53:02.269Z pid=27690 tid=xy class=ActionMailer::MailDeliveryJob jid=c932a979011a9660db5973a6 elapsed=0.078 INFO: fail
2022-06-12T23:53:02.269Z pid=27690 tid=xy WARN: {"context":"Job raised exception","job":{"retry":true,"queue":"default","class":"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper","wrapped":"ActionMailer::MailDeliveryJob","args":[{"job_class":"ActionMailer::MailDeliveryJob","job_id":"a3a62404-2f95-4b81-b223-51e6ba77f6ae","provider_job_id":null,"queue_name":"default","priority":null,"arguments":["Pay::UserMailer","receipt","deliver_now",{"params":{"pay_customer":{"_aj_globalid":"gid://gozilly/Pay::Customer/7"},"charge":{"_aj_globalid":"gid://gozilly/Pay::Charge/22"},"_aj_symbol_keys":["pay_customer","charge"]},"args":[],"_aj_ruby2_keywords":["params","args"]}],"executions":0,"exception_executions":{},"locale":"en","timezone":"Pacific Time (US & Canada)","enqueued_at":"2022-06-12T23:53:02Z"}],"jid":"c932a979011a9660db5973a6","created_at":1655077982.189643,"enqueued_at":1655077982.189813},"jobstr":"{\"retry\":true,\"queue\":\"default\",\"class\":\"ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper\",\"wrapped\":\"ActionMailer::MailDeliveryJob\",\"args\":[{\"job_class\":\"ActionMailer::MailDeliveryJob\",\"job_id\":\"a3a62404-2f95-4b81-b223-51e6ba77f6ae\",\"provider_job_id\":null,\"queue_name\":\"default\",\"priority\":null,\"arguments\":[\"Pay::UserMailer\",\"receipt\",\"deliver_now\",{\"params\":{\"pay_customer\":{\"_aj_globalid\":\"gid://gozilly/Pay::Customer/7\"},\"charge\":{\"_aj_globalid\":\"gid://gozilly/Pay::Charge/22\"},\"_aj_symbol_keys\":[\"pay_customer\",\"charge\"]},\"args\":[],\"_aj_ruby2_keywords\":[\"params\",\"args\"]}],\"executions\":0,\"exception_executions\":{},\"locale\":\"en\",\"timezone\":\"Pacific Time (US & Canada)\",\"enqueued_at\":\"2022-06-12T23:53:02Z\"}],\"jid\":\"c932a979011a9660db5973a6\",\"created_at\":1655077982.189643,\"enqueued_at\":1655077982.189813}"}
2022-06-12T23:53:02.269Z pid=27690 tid=xy WARN: KeyError: key not found: :details
2022-06-12T23:53:02.269Z pid=27690 tid=xy WARN: /Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/receipts-2.0.0/lib/receipts/base.rb:23:in `fetch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/receipts-2.0.0/lib/receipts/base.rb:23:in `generate_from'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/receipts-2.0.0/lib/receipts/base.rb:15:in `initialize'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/pay-3.0.24/lib/pay/receipts.rb:39:in `new'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/pay-3.0.24/lib/pay/receipts.rb:39:in `receipt_pdf'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/pay-3.0.24/lib/pay/receipts.rb:13:in `receipt'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/pay-3.0.24/app/mailers/pay/user_mailer.rb:5:in `receipt'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionpack-7.0.2.4/lib/abstract_controller/base.rb:214:in `process_action'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionpack-7.0.2.4/lib/abstract_controller/callbacks.rb:234:in `block in process_action'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:118:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actiontext-7.0.2.4/lib/action_text/rendering.rb:20:in `with_renderer'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actiontext-7.0.2.4/lib/action_text/engine.rb:69:in `block (4 levels) in <class:Engine>'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `instance_exec'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:138:in `run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionpack-7.0.2.4/lib/abstract_controller/callbacks.rb:233:in `process_action'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionpack-7.0.2.4/lib/abstract_controller/base.rb:151:in `process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/rescuable.rb:25:in `block in process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/rescuable.rb:17:in `handle_exceptions'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/rescuable.rb:24:in `process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionview-7.0.2.4/lib/action_view/rendering.rb:39:in `process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/base.rb:623:in `block in process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications.rb:206:in `block in instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications.rb:206:in `instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/base.rb:622:in `process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/parameterized.rb:136:in `block in processed_mailer'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/parameterized.rb:134:in `tap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/parameterized.rb:134:in `processed_mailer'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/message_delivery.rb:118:in `deliver_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/actionmailer-7.0.2.4/lib/action_mailer/mail_delivery_job.rb:23:in `perform'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/execution.rb:59:in `block in _perform_job'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:118:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/i18n-1.10.0/lib/i18n.rb:328:in `with_locale'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/translation.rb:9:in `block (2 levels) in <module:Translation>'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `instance_exec'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/core_ext/time/zones.rb:66:in `use_zone'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/timezones.rb:9:in `block (2 levels) in <module:Timezones>'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `instance_exec'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:138:in `run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/execution.rb:58:in `_perform_job'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/instrumentation.rb:20:in `_perform_job'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/execution.rb:46:in `perform_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/instrumentation.rb:14:in `block in perform_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/instrumentation.rb:25:in `block in instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications.rb:206:in `block in instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/notifications.rb:206:in `instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/instrumentation.rb:35:in `instrument'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/instrumentation.rb:14:in `perform_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/logging.rb:18:in `block in perform_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/tagged_logging.rb:99:in `block in tagged'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/tagged_logging.rb:37:in `tagged'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/tagged_logging.rb:99:in `tagged'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/logging.rb:25:in `tag_logger'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/logging.rb:18:in `perform_now'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/execution.rb:24:in `block in execute'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:118:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/railtie.rb:54:in `block (4 levels) in <class:Railtie>'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/execution_wrapper.rb:92:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/reloader.rb:72:in `block in wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/execution_wrapper.rb:88:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/reloader.rb:71:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/railtie.rb:53:in `block (3 levels) in <class:Railtie>'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `instance_exec'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:127:in `block in run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/callbacks.rb:138:in `run_callbacks'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/execution.rb:22:in `execute'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activejob-7.0.2.4/lib/active_job/queue_adapters/sidekiq_adapter.rb:42:in `perform'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:196:in `execute_job'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:164:in `block (2 levels) in process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/middleware/chain.rb:133:in `invoke'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:163:in `block in process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:136:in `block (6 levels) in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/job_retry.rb:114:in `local'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:135:in `block (5 levels) in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/rails.rb:14:in `block in call'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/execution_wrapper.rb:92:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/reloader.rb:72:in `block in wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/execution_wrapper.rb:92:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/activesupport-7.0.2.4/lib/active_support/reloader.rb:71:in `wrap'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/rails.rb:13:in `call'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:131:in `block (4 levels) in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:257:in `stats'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:126:in `block (3 levels) in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/job_logger.rb:13:in `call'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:125:in `block (2 levels) in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/job_retry.rb:81:in `global'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:124:in `block in dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/job_logger.rb:39:in `prepare'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:123:in `dispatch'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:162:in `process'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:78:in `process_one'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/processor.rb:68:in `run'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/util.rb:56:in `watchdog'
/Users/esotericcoder/.rvm/gems/ruby-2.7.2/gems/sidekiq-6.4.2/lib/sidekiq/util.rb:65:in `block in safe_thread'`

**Reply by excid3** [reply]
You can't use Receipts 2.0 with Pay 3.0. Either upgrade to Pay 4 or revert to Receipts 1.x

**Reply by EsotericCoder** [reply]
Thank you so much!! Everything works well now. Appreciate all the work to simplify payments!
